src/tvheadend/Subscription.cpp

```diff
@@ -0,0 +1,305 @@
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, write to
+ *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ *  http://www.gnu.org/copyleft/gpl.html
+ *
+ */
+
+#include "Subscription.h"
+#include "../Tvheadend.h"
+
+using namespace PLATFORM;
+
+Subscription::Subscription(CHTSPConnection &conn) :
+  m_channelId(0),
+  m_speed(1000),
+  m_weight(SUBSCRIPTION_WEIGHT_NORMAL),
+  m_id(0),
+  m_state(SUBSCRIPTION_STOPPED),
+  m_conn(conn)
+{
+}
+
+bool Subscription::IsActive() const
+{
+  CLockObject lock(m_mutex);
+  return (GetState() != SUBSCRIPTION_STOPPED);
+}
+
+uint32_t Subscription::GetId() const
+{
+  CLockObject lock(m_mutex);
+  return m_id;
+}
+
+void Subscription::SetId(uint32_t id)
+{
+  CLockObject lock(m_mutex);
+  m_id = id;
+}
+
+uint32_t Subscription::GetChannelId() const
+{
+  CLockObject lock(m_mutex);
+  return m_channelId;
+}
+
+void Subscription::SetChannelId(uint32_t id)
+{
+  CLockObject lock(m_mutex);
+  m_channelId = id;
+}
+
+uint32_t Subscription::GetWeight() const
+{
+  CLockObject lock(m_mutex);
+  return m_weight;
+}
+
+void Subscription::SetWeight(uint32_t weight)
+{
+  CLockObject lock(m_mutex);
+  m_weight = weight;
+}
+
+int32_t Subscription::GetSpeed() const
+{
+  CLockObject lock(m_mutex);
+  return m_speed;
+}
+
+void Subscription::SetSpeed(int32_t speed)
+{
+  CLockObject lock(m_mutex);
+  m_speed = speed;
+}
+
+eSubsriptionState Subscription::GetState() const
+{
+  CLockObject lock(m_mutex);
+  return m_state;
+}
+
+void Subscription::SetState(eSubsriptionState state)
+{
+  CLockObject lock(m_mutex);
+  m_state = state;
+}
+
+void Subscription::SendSubscribe(uint32_t channelId, uint32_t weight, bool restart)
+{
+  /* We do want to change anything when restarting a subscription */
+  if (!restart)
+  {
+    SetChannelId(channelId);
+    SetWeight(weight);
+    SetId(GetNextId());
+    SetSpeed(1000); //set back to normal
+  }
+
+  /* Build message */
+  htsmsg_t *m = htsmsg_create_map();
+  htsmsg_add_s32(m, "channelId",       GetChannelId());
+  htsmsg_add_u32(m, "subscriptionId",  GetId());
+  htsmsg_add_u32(m, "weight",          GetWeight());
+  htsmsg_add_u32(m, "timeshiftPeriod", (uint32_t)~0);
+  htsmsg_add_u32(m, "normts",          1);
+  htsmsg_add_u32(m, "queueDepth",      2000000);
+  tvhdebug("demux subscribe to %d",    GetChannelId());
+
+  /* Send and Wait for response */
+  if (restart)
```

Comment:
You have to follow the chain all the way up, usually when a lock is seemingly missing it's because the same lock is used further up. Or it could be a bug of course (wouldn't be the first time we screwed up the connection locking).